bgpd/bgp_route.c

```diff
@@ -6749,39 +6670,83 @@ static void set_clearing_resume_info(struct bgp_clearing_info *cinfo,
 
 /*
  * Helper to establish position in a table, possibly using "resume" info stored
- * during an iteration
+ * during an iteration. Special logic helps navigate the outer/inner tables used
+ * for the VPN safis.
  */
 static struct bgp_dest *clearing_dest_helper(struct bgp_table *table,
 					     struct bgp_clearing_info *cinfo,
 					     bool inner_p)
 {
-	struct bgp_dest *dest;
+	struct bgp_dest *dest = NULL;
 	const struct prefix *pfx;
+	bool outer_level_p;
+	char buf[PREFIX_STRLEN];
 
-	/* Iterate at start of table, or resume using inner or outer prefix */
+	/* Is this a call for the outer/RD table of a VPN safi? */
+	outer_level_p = (CHECK_FLAG(cinfo->flags, BGP_CLEARING_INFO_FLAG_INNER) && !inner_p);
+
+	/* Iterate at start of table, or resume using "inner" or "outer" (RD) prefix */
 	dest = bgp_table_top(table);
 
 	if (CHECK_FLAG(cinfo->flags, BGP_CLEARING_INFO_FLAG_RESUME)) {
+		/* We're resuming an in-progress iteration. There are a couple of cases:
+		 * 1. we're resuming in an non-VPN afi/safi; we'll look to locate the
+		 *    closest prefix _after_ the prefix we saved.
+		 * 2. if we're in a VPN safi this will be called twice: first to locate
+		 *    the RD in the "outer" table, and then a second time to locate
+		 *    the closest prefix _after_ the "inner" prefix within the RD table.
+		 */
 		pfx = NULL;
-		if (inner_p) {
-			if (CHECK_FLAG(cinfo->flags, BGP_CLEARING_INFO_FLAG_INNER))
-				pfx = &(cinfo->inner_pfx);
-		} else {
+		if (inner_p && CHECK_FLAG(cinfo->flags, BGP_CLEARING_INFO_FLAG_INNER))
+			pfx = &(cinfo->inner_pfx);
+		else
 			pfx = &(cinfo->last_pfx);
-		}
 
 		if (pfx) {
-			dest = bgp_node_match(table, pfx);
-			if (dest) {
-				/* if 'dest' matches or precedes the 'last' prefix
-				 * visited, then advance.
+			if (outer_level_p) {
+				/* Outer/RD table: use current RD if possible, advance
+				 * otherwise
 				 */
-				while (dest && (prefix_cmp(&(dest->rn->p), pfx) <= 0))
-					dest = bgp_route_next(dest);
+				prefix_rd2str((struct prefix_rd *)pfx, buf, sizeof(buf),
```

Comment:
Maybe we should call prefix_rd2str() only when debug is enabled?

Reply:
yes, sure, I'll push that change